File: src/main/java/org/railwaystations/rsapi/adapter/db/JdbiConfiguration.java
package org.railwaystations.rsapi.adapter.db;

import org.jdbi.v3.core.Jdbi;
import org.jdbi.v3.core.h2.H2DatabasePlugin;
import org.jdbi.v3.sqlobject.SqlObjectPlugin;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy;

import javax.sql.DataSource;

@Configuration
public class JdbiConfiguration {


    @Bean
    public Jdbi jdbi(final DataSource ds) {
        final TransactionAwareDataSourceProxy proxy = new TransactionAwareDataSourceProxy(ds);
        return Jdbi.create(proxy)
                .installPlugin(new H2DatabasePlugin())
                .installPlugin(new SqlObjectPlugin());
    }

    @Bean
    public CountryDao countryDao(final Jdbi jdbi) {
        return jdbi.onDemand(CountryDao.class);
    }

    @Bean
    public InboxDao inboxDao(final Jdbi jdbi) {
        return jdbi.onDemand(InboxDao.class);
    }

    @Bean
    public PhotoDao photoDao(final Jdbi jdbi) {
        return jdbi.onDemand(PhotoDao.class);
    }

    @Bean
    public StationDao stationDao(final Jdbi jdbi) {
        return jdbi.onDemand(StationDao.class);
    }

    @Bean
    public UserDao userDao(final Jdbi jdbi) {
        return jdbi.onDemand(UserDao.class);
    }

}

File: src/main/kotlin/org/railwaystations/rsapi/adapter/db/JdbiConfiguration.kt
package org.railwaystations.rsapi.adapter.db

import org.jdbi.v3.core.Jdbi
import org.jdbi.v3.core.kotlin.KotlinPlugin
import org.jdbi.v3.sqlobject.SqlObjectPlugin
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy
import javax.sql.DataSource

@Configuration
class JdbiConfiguration {
    @Bean
    fun jdbi(ds: DataSource): Jdbi {
        val proxy = TransactionAwareDataSourceProxy(ds)
        return Jdbi.create(proxy)
            .installPlugin(KotlinPlugin())
            .installPlugin(SqlObjectPlugin())
    }

    @Bean
    fun countryDao(jdbi: Jdbi): CountryDao {
        return jdbi.onDemand(CountryDao::class.java)
    }

    @Bean
    fun inboxDao(jdbi: Jdbi): InboxDao {
        return jdbi.onDemand(InboxDao::class.java)
    }

    @Bean
    fun photoDao(jdbi: Jdbi): PhotoDao {
        return jdbi.onDemand(PhotoDao::class.java)
    }

    @Bean
    fun stationDao(jdbi: Jdbi): StationDao {
        return jdbi.onDemand(StationDao::class.java)
    }

    @Bean
    fun userDao(jdbi: Jdbi): UserDao {
        return jdbi.onDemand(UserDao::class.java)
    }

    @Bean
    fun oAuth2AuthorizationDao(jdbi: Jdbi): OAuth2AuthorizationDao {
        return jdbi.onDemand(OAuth2AuthorizationDao::class.java)
    }
}
